File: VB/CustomControlBinder.vb
Imports System
Imports System.Collections.Generic
Imports DevExpress.XtraEditors

Namespace CustomFontEditControl

    Friend Module CustomControlBinder

        ' lsit of initial controls
        Private ListOfInitialControls As List(Of ListBoxControl) = New List(Of ListBoxControl)()

        Public Sub AssignControls(ByVal InitialControl As ListBoxControl, ByVal DependControl As TextEdit, ByVal ChangeDependControlValueOnLeaveFocus As Boolean)
            If InitialControl IsNot Nothing AndAlso DependControl IsNot Nothing Then
                InitialControl.Tag = DependControl
                DependControl.Tag = InitialControl
                Call ListOfInitialControls.Add(InitialControl)
                AddHandler DependControl.Properties.EditValueChanging, New DevExpress.XtraEditors.Controls.ChangingEventHandler(AddressOf TextEdit_EditValueChanging)
                If ChangeDependControlValueOnLeaveFocus Then AddHandler DependControl.Properties.Leave, New EventHandler(AddressOf TextEdit_Leave)
                AddHandler DependControl.Properties.KeyDown, New Windows.Forms.KeyEventHandler(AddressOf TextEdit_KeyDown)
                AddHandler InitialControl.SelectedIndexChanged, New EventHandler(AddressOf ListBoxControl_SelectedIndexChanged)
            End If
        End Sub

        Private Sub ListBoxControl_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim currentListBox As ListBoxControl = TryCast(sender, ListBoxControl)
            Dim currentTextEdit As TextEdit = TryCast(currentListBox.Tag, TextEdit)
            If currentListBox.Focused Then
                currentTextEdit.EditValue = currentListBox.SelectedValue
                RaiseEvent TextValueChanged(currentTextEdit, New EventArgs())
            End If
        End Sub

        Private Sub TextEdit_KeyDown(ByVal sender As Object, ByVal e As Windows.Forms.KeyEventArgs)
            Dim currentTextEdit As TextEdit = TryCast(sender, TextEdit)
            Dim currentListBox As ListBoxControl = TryCast(currentTextEdit.Tag, ListBoxControl)
            If e.KeyValue = 40 Then
                currentListBox.Focus()
            End If
        End Sub

        Private Sub TextEdit_Leave(ByVal sender As Object, ByVal e As EventArgs)
            Dim currentTextEdit As TextEdit = TryCast(sender, TextEdit)
            Dim currentListBox As ListBoxControl = TryCast(currentTextEdit.Tag, ListBoxControl)
            currentTextEdit.EditValue = currentListBox.SelectedValue
            RaiseEvent TextValueChanged(currentTextEdit, New EventArgs())
        End Sub

        Private Sub TextEdit_EditValueChanging(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.ChangingEventArgs)
            Dim currentTextEdit As TextEdit = TryCast(sender, TextEdit)
            Dim currentListBox As ListBoxControl = TryCast(currentTextEdit.Tag, ListBoxControl)
            If e.NewValue IsNot Nothing Then
                Dim iSelectedIndex As Integer = currentListBox.FindString(e.NewValue.ToString())
                If iSelectedIndex >= 0 AndAlso iSelectedIndex < currentListBox.Items.Count Then
                    currentListBox.SelectedIndex = iSelectedIndex
                End If
            End If
        End Sub

        Friend Delegate Sub CustomControlBinderrEventHandler(ByVal sender As Object, ByVal e As EventArgs)

        Public Event TextValueChanged As CustomControlBinderrEventHandler

        Public Sub DisposeControlBinder()
            For Each currentListBox As ListBoxControl In ListOfInitialControls
                Dim currentTextEdit As TextEdit = TryCast(currentListBox.Tag, TextEdit)
                RemoveHandler currentTextEdit.Properties.EditValueChanging, New DevExpress.XtraEditors.Controls.ChangingEventHandler(AddressOf TextEdit_EditValueChanging)
                RemoveHandler currentTextEdit.Properties.Leave, New EventHandler(AddressOf TextEdit_Leave)
                RemoveHandler currentTextEdit.Properties.KeyDown, New Windows.Forms.KeyEventHandler(AddressOf TextEdit_KeyDown)
                RemoveHandler currentListBox.SelectedIndexChanged, New EventHandler(AddressOf ListBoxControl_SelectedIndexChanged)
            Next

            Call ListOfInitialControls.Clear()
        End Sub
    End Module
End Namespace

File: CS/CustomControlBinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraEditors;

namespace CustomFontEditControl
{
    static class CustomControlBinder
    {
        // lsit of initial controls
        static List<ListBoxControl> ListOfInitialControls = new List<ListBoxControl>();

        public static void AssignControls(ListBoxControl InitialControl, TextEdit DependControl, bool ChangeDependControlValueOnLeaveFocus)
        {
            if (InitialControl != null && DependControl != null)
            {
                InitialControl.Tag = DependControl; DependControl.Tag = InitialControl;
                ListOfInitialControls.Add(InitialControl);
                DependControl.Properties.EditValueChanging += new DevExpress.XtraEditors.Controls.ChangingEventHandler(TextEdit_EditValueChanging);
                if (ChangeDependControlValueOnLeaveFocus) DependControl.Properties.Leave += new EventHandler(TextEdit_Leave);
                DependControl.Properties.KeyDown += new System.Windows.Forms.KeyEventHandler(TextEdit_KeyDown);
                InitialControl.SelectedIndexChanged += new EventHandler(ListBoxControl_SelectedIndexChanged);
            }        
        }

        static void ListBoxControl_SelectedIndexChanged(object sender, EventArgs e)
        {
            ListBoxControl currentListBox = sender as ListBoxControl;
            TextEdit currentTextEdit = currentListBox.Tag as TextEdit;
            if (currentListBox.Focused)
            {
                currentTextEdit.EditValue = currentListBox.SelectedValue;
                if (TextValueChanged != null) TextValueChanged(currentTextEdit, new EventArgs());
            }
        }

        static void TextEdit_KeyDown(object sender, System.Windows.Forms.KeyEventArgs e)
        {
            TextEdit currentTextEdit = sender as TextEdit;
            ListBoxControl currentListBox = currentTextEdit.Tag as ListBoxControl;
            if (e.KeyValue == 40)
            {
                currentListBox.Focus();
            }
        }

        static void TextEdit_Leave(object sender, EventArgs e)
        {
            TextEdit currentTextEdit = sender as TextEdit;
            ListBoxControl currentListBox = currentTextEdit.Tag as ListBoxControl;
            currentTextEdit.EditValue = currentListBox.SelectedValue;
            if (TextValueChanged != null) TextValueChanged(currentTextEdit, new EventArgs());
        }

        static void TextEdit_EditValueChanging(object sender, DevExpress.XtraEditors.Controls.ChangingEventArgs e)
        {
            TextEdit currentTextEdit = sender as TextEdit;
            ListBoxControl currentListBox = currentTextEdit.Tag as ListBoxControl;
            if (e.NewValue != null)
            {
                int iSelectedIndex = currentListBox.FindString(e.NewValue.ToString());
                if (iSelectedIndex >= 0 && iSelectedIndex < currentListBox.Items.Count)
                {
                    currentListBox.SelectedIndex = iSelectedIndex;
                }
            }
        }

        internal delegate void CustomControlBinderrEventHandler(object sender, EventArgs e);
        public static event CustomControlBinderrEventHandler TextValueChanged;

        public static void DisposeControlBinder()
        {
            foreach (ListBoxControl currentListBox in ListOfInitialControls)
	        {
                TextEdit currentTextEdit = currentListBox.Tag as TextEdit;
                currentTextEdit.Properties.EditValueChanging -= new DevExpress.XtraEditors.Controls.ChangingEventHandler(TextEdit_EditValueChanging);
                currentTextEdit.Properties.Leave -= new EventHandler(TextEdit_Leave);
                currentTextEdit.Properties.KeyDown -= new System.Windows.Forms.KeyEventHandler(TextEdit_KeyDown);
                currentListBox.SelectedIndexChanged -= new EventHandler(ListBoxControl_SelectedIndexChanged);        		 
	        }
            ListOfInitialControls.Clear();
        }
    }
}
